### PR TITLE
Add Pettis-Hansen sort based on call graph data from .mibc file

### PR DESCRIPTION
**Pull request by jakobbotsch**
Can be used by specifying --methodorder PettisHansen when a .mibc file
is passed that has the necessary call graph information. By default,
create-mibc adds this information.

cc @davidwrighton @AndyAyersMS

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
Would the built-in PriorityQueue type work as a replacement for the one in this PR?

**Comment by jakobbotsch** [comment]
> Would the built-in PriorityQueue type work as a replacement for the one in this PR?

It should, I'll change to use that.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                        Debug.Assert(phEdges[edge.Key][i] == phEdges[i][edge.Key]);
+                    }
+                }
+#endif
+            }
+
+            return
+                phNodes
+                .Where(n => n.Count != 0)
+                .OrderByDescending(n => n.Count)
```
Comment (davidwrighton) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs:162]: Please add a comment here that the stability of this sort is important as it preserves the default sort order for methods which are not described by optimization data as having executed. This is important as the existing default sort order is known to be substantially better than a randomized order.

**Comment by davidwrighton** [comment]
@jakobbotsch I think we can take this for .NET 6 at this time. This work is very carefully segregated from any other work in the repo and will not negatively effect any other scenarios due to that segregation. @AndyAyersMS do you agree?

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs:167]: I added a comment and an assert for this. I did not know that this was important (and it worked by accident), so it's good that you said this.

**Comment by AndyAyersMS** [comment]
> I think we can take this for .NET 6 at this time.

Yes, I think it would be useful to have this available in .NET 6.

**Review by AndyAyersMS — APPROVED** [review]
Changes LGTM.

**Review by davidwrighton — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Infra failures look unrelated

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+#if DEBUG
+            for (int i = 0; i < phNodes.Length; i++)
+            {
+                foreach (var kvp in phEdges[i])
+                    Debug.Assert(phEdges[kvp.Key][i] == phEdges[i][kvp.Key]);
+            }
+#endif
+
+            var queue = new PriorityQueue<(int from, int to), long>();
```
Comment (stephentoub) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/PettisHansenSort/PettisHansen.cs:59]: @eiriktsarpalis, just FYI.

**Pull request merged**